File: PetHealth/app/src/main/java/pe/edu/upc/pethealth/persistence/SharedPreferencesManager.java
package pe.edu.upc.pethealth.persistence;

import android.content.Context;
import android.content.SharedPreferences;

import com.google.gson.Gson;

import pe.edu.upc.pethealth.models.User;

public class SharedPreferencesManager {
    private static final String PREFERENCES_NAME = "blcc";
    private static final String CURRENT_USER = "current_user";
    private static final String ACCESS_TOKEN = "access_token";
    private static final String COMPLETED_REGISTER = "current_register";

    private static SharedPreferencesManager self;
    private final SharedPreferences mPreferences;
    private Gson gson;

    public SharedPreferencesManager(Context context) {
        this.mPreferences =context.getSharedPreferences(PREFERENCES_NAME, Context.MODE_PRIVATE);
        gson = new Gson();
    }

    public static SharedPreferencesManager getInstance(Context context){
        if (self == null){
            self = new SharedPreferencesManager(context);
        }
        return self;
    }

    public void saveUser(String user, String access_token){
        SharedPreferences.Editor editor = mPreferences.edit();
        editor.putString(CURRENT_USER, user);
        editor.putString(ACCESS_TOKEN, access_token);
        editor.apply();
    }

    public void closeSession(){
        SharedPreferences.Editor editor = mPreferences.edit();
        editor.putString(CURRENT_USER, null);
        editor.putString(COMPLETED_REGISTER, null);
        editor.apply();
    }

    public User getUser(){
        String userString = mPreferences.getString(CURRENT_USER, "");
        Gson gson = new Gson();
        if(!userString.isEmpty()){
            User user = gson.fromJson(userString, User.class);
            return user;
        }
        else{
            return null;
        }
    }

    public boolean isUserLogged(){
        return getUser() != null;
    }

    public String getAccessToken(){
        String accessToken = mPreferences.getString(ACCESS_TOKEN, "");
        return accessToken;
    }
}

File: PetHealth/app-clinic/src/main/java/pe/edu/upc/clinic/persistance/SharedPreferencesManager.kt
package pe.edu.upc.clinic.persistance


import android.content.Context
import android.content.SharedPreferences
import android.util.Log


import com.google.gson.Gson
import com.google.gson.JsonParseException
import pe.edu.upc.lib.User
import pe.edu.upc.lib.Veterinary

class SharedPreferencesManager(context: Context) {
    private val mPreferences: SharedPreferences
    private val gson: Gson

    val user: User?
        get() {
            val userString = mPreferences.getString(CURRENT_USER, "")
            val gson = Gson()
            var user: User? = null
            if(userString!!.isEmpty()) {

            } else {
                user = gson.fromJson<User>(userString.toString(), User::class.java)
            }
            return user;
        }

    val veterinary: Veterinary?
        get() {
            val veterinaryString = mPreferences.getString(CURRENT_VETERINARY, "")
            val gson = Gson()
            var veterinary: Veterinary? = null
            if (veterinaryString!!.isEmpty()) {

            } else {
                veterinary = gson.fromJson<Veterinary>(veterinaryString.toString(), Veterinary::class.java)
            }
            return veterinary
        }

    val isUserLogged: Boolean
        get() = user != null

    val accessToken: String?
        get() = mPreferences.getString(ACCESS_TOKEN, "")

    init {
        this.mPreferences = context.getSharedPreferences(PREFERENCES_NAME, Context.MODE_PRIVATE)
        gson = Gson()
    }

    fun saveUser(user: String, veterinary: String, access_token: String) {
        val editor = mPreferences.edit()
        editor.putString(CURRENT_USER, user)
        editor.putString(CURRENT_VETERINARY, veterinary)
        editor.putString(ACCESS_TOKEN, access_token)
        editor.apply()
    }

    fun closeSession() {
        val editor = mPreferences.edit()
        editor.putString(CURRENT_USER, null)
        editor.putString(CURRENT_VETERINARY, null)
        editor.putString(COMPLETED_REGISTER, null)
        editor.apply()
    }

    companion object {
        private val PREFERENCES_NAME = "blcc"
        private val CURRENT_USER = "current_user"
        private val CURRENT_VETERINARY = "current_veterinary"
        private val ACCESS_TOKEN = "access_token"
        private val COMPLETED_REGISTER = "current_register"

        private var self: SharedPreferencesManager? = null

        fun getInstance(context: Context): SharedPreferencesManager {
            if (self == null) {
                self = SharedPreferencesManager(context)
            }
            return self as SharedPreferencesManager
        }
    }
}